File: samples/basicECCClient/src/uk/ac/soton/itinnovation/experimedia/arch/ecc/samples/basicECCClient/ECCClientController.java
/////////////////////////////////////////////////////////////////////////
//
// © University of Southampton IT Innovation Centre, 2012
//
// Copyright in this software belongs to University of Southampton
// IT Innovation Centre of Gamma House, Enterprise Road, 
// Chilworth Science Park, Southampton, SO16 7NS, UK.
//
// This software may not be used, sold, licensed, transferred, copied
// or reproduced in whole or in part in any manner or form or in or
// on any media by any person other than in accordance with the terms
// of the Licence Agreement supplied with the software, or otherwise
// without the prior written consent of the copyright owners.
//
// This software is distributed WITHOUT ANY WARRANTY, without even the
// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
// PURPOSE, except where stated in the Licence Agreement supplied with
// the software.
//
//      Created By :            Simon Crowle
//      Created Date :          15-Aug-2012
//      Created for Project :   EXPERIMEDIA
//
/////////////////////////////////////////////////////////////////////////

package uk.ac.soton.itinnovation.experimedia.arch.ecc.samples.basicECCClient;

import uk.ac.soton.itinnovation.experimedia.arch.ecc.samples.shared.*;
import uk.ac.soton.itinnovation.experimedia.arch.ecc.amqpAPI.impl.amqp.*;
import uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.metrics.*;
import uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.monitor.*;
import uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.experiment.Experiment;

import org.apache.log4j.Logger;

import java.util.*;
import javax.swing.JOptionPane;




public class ECCClientController implements EMIAdapterListener,
                                            ECCClientViewListener
{
    private final Logger clientLogger = Logger.getLogger( ECCClientController.class );

    private AMQPBasicChannel   amqpChannel;
    private EMInterfaceAdapter emiAdapter;
    private ECCClientView      clientView;
    private String             clientName;

    private Entity                        entityBeingObserved;
    private Attribute                     entityAttribute;
    private HashMap<UUID,MetricGenerator> metricGenerators;

    private Measurement firstMeasurement;
    private Measurement currentMeasurement;
    
    private HashMap<UUID, Report> pendingPushReports;
    private HashMap<UUID, Report> pendingPullReports;



    public ECCClientController()
    {
        metricGenerators   = new HashMap<UUID,MetricGenerator>();
        pendingPushReports = new HashMap<UUID, Report>();
        pendingPullReports = new HashMap<UUID, Report>();
    }

    public void start( String rabbitServerIP,
                       UUID expMonitorID,
                       UUID clientID ) throws Exception
    {
        if ( rabbitServerIP != null &&
             expMonitorID   != null &&
             clientID       != null )
        {
            clientLogger.info( "Trying to connect to Rabbit server on " + rabbitServerIP );

            // Create connection to Rabbit server ------------------------------------
            AMQPConnectionFactory amqpFactory = new AMQPConnectionFactory();
            amqpFactory.setAMQPHostIPAddress( rabbitServerIP );
            try
            {
                amqpFactory.connectToAMQPHost();
                amqpChannel = amqpFactory.createNewChannel();
            }
            catch (Exception e ) 
            {
                clientLogger.error( "Could not connect to Rabbit server" );
                throw e; 
            }

            // Set up a simple view --------------------------------------------------
            Date date = new Date();
            clientName = date.toString();
            clientView = new ECCClientView( clientName, this );
            clientView.setVisible( true );

            // Create EM interface adapter, listen to it...
            emiAdapter = new EMInterfaceAdapter( this );

            // ... and try registering with the EM.

            try { emiAdapter.registerWithEM( clientName,
                                            amqpChannel, 
                                            expMonitorID, clientID ); }
            catch ( Exception e ) 
            {
                clientLogger.error( "Could not attempt registration with EM" );
                throw e; 
            }
        }
    }

    // EMIAdapterListener --------------------------------------------------------
    @Override
    public void onEMConnectionResult( boolean connected, Experiment expInfo )
    {
        if ( connected )
        {
          clientView.setStatus( "Connected to EM" );
          clientView.addLogMessage( "Linked to experiment: " + expInfo.getName() );
        }
        else
          clientView.setStatus( "Refused connection to EM" );
    }

    @Override
    public void onEMDeregistration( String reason )
    {
        clientView.addLogMessage( "Got disconnected from EM: " + reason );
        
        try
        { emiAdapter.disconnectFromEM(); }
        catch ( Exception e )
        { clientLogger.error( "Had problems disconnecting from EM: " + e.getMessage() ); }
        
        // Apologise to the user
        JOptionPane.showMessageDialog( null, "ECC disconnected this client: " + reason );
        
        clientView.dispose();
        System.exit(0);
    }
    
    @Override
    public void onDescribeSupportedPhases( EnumSet<EMPhase> phasesOUT )
    {
        // We're going to support all phases (although we won't do much in some of them)
        // ... we MUST support the discovery phase by default, but don't need to include
        phasesOUT.add( EMPhase.eEMSetUpMetricGenerators );
        phasesOUT.add( EMPhase.eEMLiveMonitoring );
        phasesOUT.add( EMPhase.eEMPostMonitoringReport );
        phasesOUT.add( EMPhase.eEMTearDown );
    }
    
    @Override
    public void onDescribePushPullBehaviours( Boolean[] pushPullOUT )
    {
        // We're going to support both push and pull
        pushPullOUT[0] = true;
        pushPullOUT[1] = true;
    }
    
    @Override
    public void onPopulateMetricGeneratorInfo()
    {
        clientView.setStatus( "Sending metric gen info to EM" );

        // Create a new entity to be observed (this Java VM)
        entityBeingObserved = new Entity();
        entityBeingObserved.setName( "EM Client host" );

        // Create an attribute to observe
        entityAttribute = new Attribute();
        entityAttribute.setName( "Client RAM usage" );
        entityAttribute.setDescription( "Very simple measurement of total bytes used" );
        entityAttribute.setEntityUUID( entityBeingObserved.getUUID() );
        entityBeingObserved.addAttribute( entityAttribute );

        // Create a single metric generator that will represent this metric data generation
        MetricGenerator metricGen = new MetricGenerator();
        metricGen.setName( "MGEN " + clientName );
        metricGen.setDescription( "Metric generator demonstration" );
        metricGen.addEntity( entityBeingObserved );

        // Add our generator to a collection - obviously we only actually have one generator here though
        metricGenerators.put( metricGen.getUUID(), metricGen );

        // Create a single group which will contain...
        MetricGroup mg = new MetricGroup();
        mg.setName( "Demo group" );
        mg.setDescription( "A single group to contain metrics" );
        mg.setMetricGeneratorUUID( metricGen.getUUID() );
        metricGen.addMetricGroup( mg );

        // ... a single MeasurementSet (representing the measures for the attibute)
        MeasurementSet ms = new MeasurementSet();
        ms.setMetricGroupUUID( mg.getUUID() );
        ms.setAttributeUUID( entityAttribute.getUUID() ); // Link the measurement set to the attribute here
        mg.addMeasurementSets( ms );                              

        // Define the metric for this MeasurementSet
        Metric memMetric = new Metric();
        memMetric.setMetricType( MetricType.RATIO );
        memMetric.setUnit( new Unit("bytes") );
        ms.setMetric( memMetric );

        clientView.addLogMessage( "Discovered generator: " + metricGen.getName() );

        // Ready our metric generator for the EM
        HashSet mgSet = new HashSet<MetricGenerator>();
        mgSet.addAll( metricGenerators.values() );
        emiAdapter.setMetricGenerators( mgSet );
    }

    @Override
    public void onDiscoveryTimeOut()
    { clientView.addLogMessage( "Got discovery time-out message" ); }
    
    @Override
    public void onSetupMetricGenerator( UUID genID, Boolean[] resultOUT )
    {
        clientView.setStatus( "Setting up generators" );

        // Just signal that the metric generator is ready
        resultOUT[0] = true;

        clientView.addLogMessage( "Completed generator set-up" );
    }
    
    @Override
    public void onSetupTimeOut( UUID metricGeneratorID )
    { clientView.addLogMessage( "Got set-up time-out message" ); }
    
    @Override
    public void onLiveMonitoringStarted()
    {
        clientView.addLogMessage( "ECC has started Live Monitoring process" );
    }

    @Override
    public void onStartPushingMetricData()
    {
        // Allow the human user to manually push some data
        clientView.addLogMessage( "Enabling metric push" );
        clientView.enablePush( true );
    }

    @Override
    public void onPushReportReceived( UUID reportID )
    {
        // We'll use this report with the 'MiniEDM' later, but for now...
        pendingPushReports.remove( reportID );
      
        // Got the last push, so allow another manual push
        clientView.enablePush( true );
    }
    
    @Override
    public void onPullReportReceived( UUID reportID )
    {
        // We'll use this report with the 'MiniEDM' later, but for now...
        pendingPullReports.remove( reportID );
    }
    
    @Override
    public void onPullMetricTimeOut( UUID measurementSetID )
    { clientView.addLogMessage( "Got live pull time-out message" ); }

    @Override
    public void onStopPushingMetricData()
    {
        // Stop manual pushing of data by the human user
        clientView.addLogMessage( "Disabling metric push" );
        clientView.enablePush( false );
    }

    @Override
    /*
    * Note that 'reportOut' is an OUT parameter provided by the adapter
    */
    public void onPullMetric( UUID measurementSetID, Report reportOut )
    {
        // Create an empty instance of our measurement set
        MeasurementSet sampleSet = createMeasurementSetEmptySample();

        // Add a snapshot measurement to it
        snapshotMeasurement();
        sampleSet.addMeasurement( currentMeasurement );

        // Set up report (has only a single measure)
        Date date = new Date();
        reportOut.setReportDate( date );
        reportOut.setFromDate( date );
        reportOut.setToDate( date );
        reportOut.setMeasurementSet( sampleSet );
        reportOut.setNumberOfMeasurements( 1 );
        
        // Store for confirmation of pull later
        pendingPullReports.put( reportOut.getUUID(), reportOut );
    }

    @Override
    public void onPullingStopped()
    {
        clientView.addLogMessage( "ECC has stopped pulling" );
    }
    
    @Override
    /*
    * Note that the summaryOUT parameter is an OUT parameter supplied by the
    * adapter
    */
    public void onPopulateSummaryReport( EMPostReportSummary summaryOUT )
    {
        // We've only got one MeasurementSet so we'll create a demo summary report
        // and just two measurements.. so we'll use these to create a demo summary

        // If we don't have any measurements, make some up!
        if ( firstMeasurement   ==  null ) snapshotMeasurement();
        if ( currentMeasurement ==  null ) snapshotMeasurement();

        // Create a new report for this summary
        Report report = new Report();
        report.setReportDate( new Date() );
        report.setFromDate( firstMeasurement.getTimeStamp() );
        report.setToDate( currentMeasurement.getTimeStamp() );
        report.setNumberOfMeasurements( 2 );

        // We've only got one of each...
        MetricGenerator mGen = metricGenerators.values().iterator().next();
        MetricGroup mGroup = mGen.getMetricGroups().iterator().next();
        MeasurementSet mSet = mGroup.getMeasurementSets().iterator().next();

        // Add our one MeasurementSet data to the report and add that to the summary report
        report.setMeasurementSet( mSet );
        summaryOUT.addReport( report );
    }

    @Override
    public void onPopulateDataBatch( EMDataBatch batchOUT )
    {
        // We've only stored the first and the last measurements of a single
        // MeasurementSet, so just send that
        MeasurementSet ms = createMeasurementSetEmptySample();
        ms.addMeasurement( firstMeasurement );
        ms.addMeasurement( currentMeasurement );
        
        Report batchRep = new Report();
        batchRep.setFromDate( firstMeasurement.getTimeStamp() );
        batchRep.setToDate( currentMeasurement.getTimeStamp() );
        batchRep.setMeasurementSet( ms );
        batchRep.setNumberOfMeasurements( 2 );
        
        batchOUT.setBatchReport( batchRep );
    }
    
    @Override
    public void onReportBatchTimeOut( UUID batchID )
    { clientView.addLogMessage( "Got post-report time-out message" ); }

    @Override
    public void onGetTearDownResult( Boolean[] resultOUT )
    {
        clientView.setStatus( "Tearing down" );
        clientView.addLogMessage( "Tearing down metric generators" );

        // Signal we've successfully torn-down
        resultOUT[0] = true;
    }

    // EMClientViewListener ------------------------------------------------------
    @Override
    public void onPushDataClicked()
    {
        // Create an empty instance of our measurement set
        MeasurementSet sampleSet = createMeasurementSetEmptySample();

        // Take a current measurement
        snapshotMeasurement();
        sampleSet.addMeasurement( currentMeasurement );

        // Set up report (has only a single measure)
        Date date           = new Date();
        Report sampleReport = new Report();

        sampleReport.setMeasurementSet( sampleSet );
        sampleReport.setReportDate(date );
        sampleReport.setFromDate( date );
        sampleReport.setToDate( date );
        sampleReport.setMeasurementSet( sampleSet );
        sampleReport.setNumberOfMeasurements( 1 );

        // ... and store for confirmation of push, then report!
        pendingPushReports.put( sampleReport.getUUID(), sampleReport );
        emiAdapter.pushMetric( sampleReport );
    }
    
    @Override
    public void onClientViewClosed()
    {
        // Need to notify that we're leaving...
        try { emiAdapter.disconnectFromEM(); }
        catch ( Exception e )
        { clientLogger.error( "Could not cleanly disconnect from EM:\n" + e.getMessage() ); }
    }
    
    @Override
    public void onTearDownTimeOut()
    { clientView.addLogMessage( "Got tear-down time-out message" ); }

    // Private method ------------------------------------------------------------
    private MeasurementSet createMeasurementSetEmptySample()
    {
        // Get our only metric generator
        MetricGenerator metGen = metricGenerators.values().iterator().next();
        metGen.getMetricGroups().iterator().next();

        // Get our only metric group
        MetricGroup mg = metGen.getMetricGroups().iterator().next();
        MeasurementSet currentMS = mg.getMeasurementSets().iterator().next();

        return new MeasurementSet( currentMS, false );
    }

    private void snapshotMeasurement()
    {
        // Just take a very rough measurement
        Runtime rt = Runtime.getRuntime();
        String memVal = Long.toString( rt.totalMemory() - rt.freeMemory() );

        // Get the (single) measurement set for this snapshot (just to get the ID actually)
        MeasurementSet snapshotMS = createMeasurementSetEmptySample();

        // Update the latest measurement
        currentMeasurement = new Measurement();
        currentMeasurement.setMeasurementSetUUID( snapshotMS.getID() );
        currentMeasurement.setTimeStamp( new Date() );
        currentMeasurement.setValue( memVal );

        // Store this if it is the first ever measurement
        if ( firstMeasurement == null ) firstMeasurement = currentMeasurement;

        clientView.addLogMessage( "Memory measurement (bytes): " + memVal );
    }
}

File: samples/basicCPPClient/ECCClientController.cpp
/////////////////////////////////////////////////////////////////////////
//
// © University of Southampton IT Innovation Centre, 2013
//
// Copyright in this software belongs to University of Southampton
// IT Innovation Centre of Gamma House, Enterprise Road, 
// Chilworth Science Park, Southampton, SO16 7NS, UK.
//
// This software may not be used, sold, licensed, transferred, copied
// or reproduced in whole or in part in any manner or form or in or
// on any media by any person other than in accordance with the terms
// of the Licence Agreement supplied with the software, or otherwise
// without the prior written consent of the copyright owners.
//
// This software is distributed WITHOUT ANY WARRANTY, without even the
// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
// PURPOSE, except where stated in the Licence Agreement supplied with
// the software.
//
//      Created By :            Simon Crowle
//      Created Date :          08-Jul-2013
//      Created for Project :   EXPERIMEDIA
//
/////////////////////////////////////////////////////////////////////////

#include "StdAfx.h"

#include "ECCClientController.h"

#include "MetricHelper.h"

#include <iostream>

using namespace ecc_amqpAPI_impl;
using namespace ecc_commonDataModel;

using namespace boost;
using namespace std;



ECCClientController::ECCClientController()
  : connectedToECC(false)
{
}

ECCClientController::~ECCClientController()
{
    emiAdapter = NULL;
    
    inAMQPChannel = NULL;
    outAMQPChannel = NULL;
}

void ECCClientController::start( const String& rabbitServerIP,
                                 const UUID&   expMonitorID,
                                 const UUID&   clientID )
{
  if ( !rabbitServerIP.empty() && !connectedToECC )
  {
    wcout << L"Attempting to connect to Rabbit" << endl;

    amqpFactory = AMQPConnectionFactory::ptr_t( new AMQPConnectionFactory() );

    try
    {
      // Set up AMQP factory to connect to the RabbitMQ server
      amqpFactory->setAMQPHostIPAddress( rabbitServerIP );
      amqpFactory->connectToAMQPHost();
      inAMQPChannel  = amqpFactory->createNewChannel();
      outAMQPChannel = amqpFactory->createNewChannel();

      cout << "Connected and channel created." << endl;

      // Create a name for this client (based on the time)
      TimeStamp date = getCurrentTime();
      clientName     = L"C++ Client (" + timeStampToString(date) + L")";

      // Create an adapter that will manage the 'low level' ECC communications
      emiAdapter = 
        EMInterfaceAdapter::ptr_t( new EMInterfaceAdapter( shared_from_this() ) );

      // Try registering with the ECC (see onEMConnectionResult(..) for result)
      emiAdapter->registerWithEM( clientName,
                                  inAMQPChannel, outAMQPChannel,
                                  expMonitorID,
                                  clientID );
    }
    catch ( const String& e ) 
    { wcout << L"Had problems connecting to RabbitMQ: " << e; }
  }
}

void ECCClientController::stop()
{
    if ( emiAdapter ) emiAdapter->disconnectFromEM();

    emiAdapter = NULL;

    if ( amqpFactory )
    {
        amqpFactory->closeDownConnection();
        amqpFactory = NULL;

        inAMQPChannel = NULL;
        outAMQPChannel = NULL;
    }    
}

// EMInterfaceAdapterListener ------------------------------------------------
void ECCClientController::onEMConnectionResult( const bool              connected, 
                                                const Experiment::ptr_t expInfo )
{
  wcout << L"Got connection result: " ;

  // If we successfully connected to the ECC, then display some basic info
  if ( connected )
  {
    wcout << L"CONNECTION OK" << endl << endl;
    wcout << L"Linked to experiment: " + expInfo->getExperimentID() << endl;

    connectedToECC = true;
  }
  else
  {
    wcout << L"CONNECTION REFUSED BY ECC" << endl;
    connectedToECC = false;
  }
}
    
void ECCClientController::onEMDeregistration( const String& reason )
{
    wcout << L"Got disconnection from the ECC: " << reason << endl;

    connectedToECC = false;
}
    
void ECCClientController::onDescribeSupportedPhases( EMPhaseSet& phasesOUT )
{
    // We are going to just support the live monitoring phase...
    // ... we support 'Discovery' by default

    wcout << L"Specifying phase compatibility (Live monitoring only)" << endl;

    phasesOUT.insert( ecc_commonDataModel::eEMLiveMonitoringPhase );

}

void ECCClientController::onDescribePushPullBehaviours( bool* pushPullOUT )
{
    // We're going to support only PULLing behaviour
    wcout << L"Specifying only pull behaviour" << endl;

    pushPullOUT[0] = false; // We will not PUSH data in this demo
    pushPullOUT[1] = true;  // We will response to PULL requests in this demo
}
    
void ECCClientController::onPopulateMetricGeneratorInfo()
{
    wcout << "Creating metric generator info..." << endl;

    Entity::ptr_t app = Entity::ptr_t( new Entity( createRandomUUID(),
                                                   L"Client app",
                                                   L"The thing we are observing") );

    Attribute::ptr_t alphaAttr = 
      MetricHelper::createAttribute( L"Random stream Alpha",
                                     L"A random number stream",
                                     app );

    Attribute::ptr_t betaAttr 
      = MetricHelper::createAttribute( L"Random stream Beta",
                                       L"Another random number stream",
                                       app );

    // Create our metric generator (we'll use this later during live monitoring)
    metricGenerator = MetricGenerator::ptr_t( new MetricGenerator( createRandomUUID(),
                                              L"Client metric generator",
                                              L"The only metric generator for this client" ) );
    // Associate an entity with the generator
    metricGenerator->addEntity( app );

    // Create a group of metric relating to memory
    MetricGroup::ptr_t group = 
      MetricHelper::createMetricGroup( L"Sample metric group",
                                       L"Metrics relating to the app",
                                       metricGenerator );

    // Create measurment sets for our two metrics
    MeasurementSet::ptr_t alphaSet =
        MetricHelper::createMeasurementSet( alphaAttr,
                                            RATIO,
                                            Unit::ptr_t( new Unit(L"Random point") ),
                                            group );  

    MeasurementSet::ptr_t betaSet =
        MetricHelper::createMeasurementSet( betaAttr,
                                            RATIO,
                                            Unit::ptr_t( new Unit(L"Random point") ),
                                            group );

    // Create some measurement function delegates to do the actual measuring
    MeasurementDelegate alphaDelegate = &ECCClientController::measureAlphaDelegate;
    MeasurementDelegate betaDelegate  = &ECCClientController::measureBetaDelegate;

    // Map delegates to the appropriate measurement set IDs (we'll use these when PULLED by the ECC)
    delegateMap.insert( DelegateMap::value_type( alphaSet->getID(), alphaDelegate ) );
    delegateMap.insert( DelegateMap::value_type( betaSet->getID(),  betaDelegate ) );

    // Wrap up our metric generator and send to the ECC
    MetricGenerator::Set mgSet;
    mgSet.insert( metricGenerator );

    // Send to the ECC
    emiAdapter->sendMetricGenerators( mgSet );

    wcout << L"...sent metric generators to ECC" << endl;
}
    
void ECCClientController::onDiscoveryTimeOut()
{
    /* Not implemented in this demo */
}

void ECCClientController::onSetupMetricGenerator( const UUID& metricGeneratorID, 
                                                  bool* resultOUT )
{
    /* Not implemented in this demo */
}
    
void ECCClientController::onSetupTimeOut( const UUID& metricGeneratorID )
{
    /* Not implemented in this demo */
}
    
void ECCClientController::onLiveMonitoringStarted()
{
    wcout << L"Live monitoring has started" << endl;
}

void ECCClientController::onStartPushingMetricData()
{
    wcout << L"Got command to start pushing (we said we wouldn't, so shouldn't see this)" << endl;
}
    
void ECCClientController::onPushReportReceived( const UUID& lastReportID )
{
    wcout << L"Got push confirmed (shouldn't see this)" << endl;
}
    
void ECCClientController::onStopPushingMetricData()
{
    wcout << L"Got command stop pushing (shouldn't see this)" << endl;
}
    
void ECCClientController::onPullReportReceived( const UUID& reportID )
{
    wcout << L"Got confirmed pull (report ID): " << uuidToWide(reportID) << endl;
}
    
void ECCClientController::onPullMetric( const UUID& measurementSetID, 
                                        Report::ptr_t reportOUT )
{
    wcout << L"Got pull request for measurement set: " << uuidToWide(measurementSetID) << endl;

    // Select the correct delegate to make this measurement (based on measurement set ID)
    DelegateMap::iterator delIt = delegateMap.find( measurementSetID );

    if ( delIt != delegateMap.end() )
    {
        MeasurementDelegate mDelegate = delIt->second;
      
        // Create a measurement report using the delegate call
        Report::ptr_t newReport = mDelegate( shared_from_this(), measurementSetID );
  
        // Copy the report into the OUT paramter of this method (it gets sent to the ECC)
        reportOUT->copyReport( newReport, true );
    }
}
    
void ECCClientController::onPullMetricTimeOut( const UUID& measurementSetID )
{
    /* Not implemented in this demo */
}

void ECCClientController::onPullingStopped()
{
    wcout << L"ECC has stopped pulling metrics" << endl;
}

void ECCClientController::onPopulateSummaryReport( EMPostReportSummary::ptr_t summaryOUT )
{
    /* Not implemented in this demo */
}

void ECCClientController::onPopulateDataBatch( EMDataBatch::ptr_t batchOut )
{
    /* Not implemented in this demo */
}
    
void ECCClientController::onReportBatchTimeOut( const UUID& batchID )
{
    /* Not implemented in this demo */
}

void ECCClientController::onGetTearDownResult( bool* resultOUT )
{
    /* Not implemented in this demo */
}
    
void ECCClientController::onTearDownTimeOut()
{
    /* Not implemented in this demo */
}

// Private methods ------------------------------------------------------------------

Report::ptr_t ECCClientController::measureAlphaDelegate( UUID msID )
{
    wcout << L"Creating measurement for ALPHA" << endl;

    // Just going to create a random number; your specific measurement algorithm would go here
    return createRandomMeasurement( msID );
}

Report::ptr_t ECCClientController::measureBetaDelegate( UUID msID )
{
    wcout << L"Creating measurement for BETA" << endl;

    // Just going to create a random number; your specific measurement algorithm would go here
    return createRandomMeasurement( msID );
}

Report::ptr_t ECCClientController::createRandomMeasurement( UUID msID )
{
    // Get the measurement set requested from our metric generator
    MeasurementSet::ptr_t targetSet = MetricHelper::getMeasurementSet( metricGenerator, msID );
    
    // Create a new report based on the specific measurement set
    Report::ptr_t newReport = MetricHelper::createEmptyMeasurementReport( targetSet );

    // Create random value
    int r = rand();
    r = (r >> 8);
    String value = intToString( r );

    // Encapsulate the raw measurement and insert into the report
    Measurement::ptr_t m = Measurement::ptr_t( new Measurement(value) );
    newReport->getMeasurementSet()->addMeasurement( m );
    newReport->setNumberOfMeasurements( 1 );

    return newReport;
}